File: packages/engine/src/agentgate_engine/nodes/retriever.py
"""Node 2 — pull the policies that actually bear on this action."""

from __future__ import annotations

import json
from typing import Any

from ..enrich import enrich_query
from ..policy_store import retrieve_policies
from ..state import GraphState


def to_query(state: GraphState) -> str:
    """Turn a tool call into the query we search policies with, annotated with
    the kinds of sensitive data detected in its arguments."""
    action = state["action"]
    payload = json.dumps(action.tool_args or {}, default=str)
    base = (
        f'Category {state["category"]}. Tool "{action.tool_name}" '
        f"called with arguments: {payload}"
    )
    return enrich_query(base, f"{action.tool_name} {payload}")


async def retriever_node(state: GraphState) -> dict[str, Any]:
    policies = await retrieve_policies(
        to_query(state), category=state["category"], trace=state["trace"]
    )
    return {"policies": policies}

File: packages/engine/src/nodes/retriever.ts
import { enrichQuery } from '../enrich.ts';
import { retrievePolicies } from '../policyStore.ts';
import type { GraphState } from '../state.ts';

/**
 * Turn a tool call into the query we search policies with, annotated with the
 * kinds of sensitive data detected in its arguments so that a raw payload can
 * match policy prose it shares no vocabulary with.
 */
export function toQuery(state: GraphState): string {
  const { toolName, toolArgs } = state.action;
  const payload = JSON.stringify(toolArgs ?? {});
  const base = `Category ${state.category}. Tool "${toolName}" called with arguments: ${payload}`;
  return enrichQuery(base, `${toolName} ${payload}`);
}

export async function retrieverNode(state: GraphState): Promise<Partial<GraphState>> {
  const policies = await retrievePolicies(toQuery(state), {
    category: state.category,
    trace: state.trace,
  });
  return { policies };
}
